### PR TITLE
Fix included record parameter issue with non-basic types

### LINKED ISSUE #27458: Compiler crashes when argument is not a simple basic type(With included record parameter)
**Description:**
$Title.

**Steps to reproduce:**
```ballerina
public type KeyValues record {};

function bar(*KeyValues baz) returns string {
    return "";
}

public function main() {
    string result = bar(username = [1, 2]);
}
```

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #27458 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -5680,9 +5680,10 @@ private BVarSymbol checkParameterNameForDefaultArgument(BLangIdentifier argName,
             }
         }
         if (incRecordParamAllowAdditionalFields != null) {
-            LinkedHashMap<String, BField> fields = ((BRecordType) incRecordParamAllowAdditionalFields.type).fields;
-            if (!fields.containsKey(argName.value) && types.isAssignable(expr.type,
-                                            ((BRecordType) incRecordParamAllowAdditionalFields.type).restFieldType)) {
+            BRecordType incRecordType = (BRecordType) incRecordParamAllowAdditionalFields.type;
+            BType exprType = checkExpr(expr, env);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5684]: As discussed offline, let's check if we can use the expected type for the argument here to make use of the contextual information.

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -5680,9 +5680,10 @@ private BVarSymbol checkParameterNameForDefaultArgument(BLangIdentifier argName,
             }
         }
         if (incRecordParamAllowAdditionalFields != null) {
-            LinkedHashMap<String, BField> fields = ((BRecordType) incRecordParamAllowAdditionalFields.type).fields;
-            if (!fields.containsKey(argName.value) && types.isAssignable(expr.type,
-                                            ((BRecordType) incRecordParamAllowAdditionalFields.type).restFieldType)) {
+            BRecordType incRecordType = (BRecordType) incRecordParamAllowAdditionalFields.type;
+            checkExpr(expr, env, incRecordType.restFieldType);
+            LinkedHashMap<String, BField> fields = incRecordType.fields;
+            if (!fields.containsKey(argName.value)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5686]:
```suggestion
            if (!incRecordType.fields.containsKey(argName.value)) {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_included_record_parameters.bal [hunk 3]
```diff
@@ -277,6 +283,12 @@ function testFunctionOfFunctionTypedParamWithIncludedRecordParam20() {
     assertEquality(5, val);
 }
 
+function testFunctionOfFunctionTypedParamWithIncludedRecordParam21() {
+    int[] x = [1, 2];
+    int[] val = functionOfFunctionTypedParamWithIncludedRecordParam21(a = [1, 2]);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_included_record_parameters.bal:288]: Shall we also add a test where the contextually expected type is used? e.g., `[1, 2]` where a `float[]` is expected.

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_included_record_parameters.bal:293]: Added.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Changed

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**